File: src/makefile_generator.py
""" File: generateMakefiles.py
    Version: 1.0.1
    Date: 16-01-2021 
    Author: David Villalobos
    Description: With this file you can 
    generate makefiles for build C++ projects
"""

import sys
import os

def help():
    return  """ [Generator of Makefiles for C++]
    Author: https://github.com/DavidVillalobos
    Usages:
    * Interactive console
        makefile_generator.py
    * Through arguments
        makefile_generator.py [main] [out] [name_class_1] [name_class_2] ...
        """

def version(): return 'V1.0.1'

def generate_makefile(main, classes, out, flags, targets):
    makefile = 'OBJS = ' + main + '.o ' + '.o '.join(classes) + ('.o\n' if 0 < len(classes) else '\n')
    makefile += 'SOURCE = ' + main + '.cpp ' + '.cpp '.join(classes) + ('.cpp\n' if 0 < len(classes) else '\n')
    makefile += 'HEADER = ' + '.h '.join(classes) + ('.h\n' if 0 < len(classes) else '\n')
    makefile += 'OUT = ' + out + '.exe\n'
    makefile += 'CC = g++\n'
    makefile += 'FLAGS = -c ' 
    if flags['g']: 
        makefile += '-g '  
    if flags['Wall']: 
        makefile += '-Wall'   
    makefile += '\nLFLAGS =\n\n' 
    makefile += 'all: $(OBJS)\n'
    makefile += '\t$(CC) -g $(OBJS) -o $(OUT) $(LFLAGS)\n\n'
    makefile += main + '.o: ' + main + '.cpp\n'
    makefile += '\t$(CC) $(FLAGS) ' + main + '.cpp\n\n'
    for i in classes: 
        makefile += i + '.o: ' + i + '.cpp\n\t$(CC) $(FLAGS) ' + i + '.cpp\n\n'
    if targets['clean']: 
        makefile += 'clean:\n'
        makefile += '\tdel $(OBJS) $(OUT)\n\n'
    if targets['run']: 
        makefile += 'run: $(OUT)\n'
        makefile += '\t$(OUT)\n'
    path = 'C:\\Users\\' + os.getenv('username') + '\\Desktop\\Makefile'
    with open (path, 'w') as file:
        file.write(makefile)
    file.close()
    return makefile

if __name__ == '__main__':
    output = ''
    if len(sys.argv) > 1:
        if sys.argv[1] == '--help' or sys.argv[1] == '-help' or sys.argv[1] == '-h': print(help())
        elif sys.argv[1] == '--version' or sys.argv[1] == '-version' or sys.argv[1] == '-v': print(version())
        elif len(sys.argv) > 2:
            print('Creating Makefile for c++ through arguments')
            main = sys.argv[1]
            out = sys.argv[2]
            classes = [ sys.argv[i] for i in range(3, len(sys.argv))]
            flags = {}
            flags['g'] = input('Allow debugging (Y/n):') != 'n'
            flags['Wall'] = input('Show all warnings (Y/n):') != 'n'
            targets = {}
            targets['clean'] = input('Include clean target (Y/n):') != 'n'
            targets['run'] = input('Include run target (Y/n):') != 'n'
            output = generate_makefile(main, classes, out, flags, targets)
            print("\nThe result of the makefile is as follows\n\n" + output)
        else:
            output += "Error: Not give anyone name for .exe "
    else:
        print('Creating Makefile for c++ with console interactive')
        classes = []
        main = input('Name Main: ')
        out = input('Name Out: ')
        while input('New class (Y/n):') != 'n': classes.append((input('Name: ')))
        flags = {}
        flags['g'] = input('Allow debugging (Y/n):') != 'n'
        flags['Wall'] = input('Show all warnings (Y/n):') != 'n'
        targets = {}
        targets['clean'] = input('Include clean target (Y/n):') != 'n'
        targets['run'] = input('Include run target (Y/n):') != 'n'
        output = generate_makefile(main, classes, out, flags, targets)
        print("\nThe result of the makefile is as follows\n\n" + output)

File: src/makefile_generator.cpp
/*
    File: generateMakefiles.cpp
    Version: 2.0.0
    Date: 16-03-2021 
    Author: David Villalobos
    Description: With this file you can 
    generate makefiles for build C++ projects
*/
#include<iostream>
#include<cstdio>
#include<fstream>
#include<sstream>
#include<list>

#define VERSION "GenMakeC++ V2.0.0"

// Get help like -h or -help
std::string help();

// Display a list elements like string concatenate with a separator 
std::string displayList(std::list<std::string> &list, std::string separator = " ");

// generate the targets to compile each file in list
std::string compileTargetFiles(std::list<std::string> list);

// generate the makefile according to main.cpp, classes.h out.exe, flags and library flags, clean and run target
std::string generate_makefile(std::string &main, std::list<std::string> &classes, 
std::string &out, std::string flags, std::string lflags, bool clean, bool run);

int main(int argc, char** argv){
    if (argc > 1){
        if(std::string(argv[1]) == "--help" || std::string(argv[1]) == "-help" || std::string(argv[1]) == "-h"){
            std::cout << help();
        } else if(std::string(argv[1]) == "--version" || std::string(argv[1]) == "-version" || std::string(argv[1]) == "-v"){
            std::cout << VERSION;
        } else if(argc > 2){
            std::string main = std::string(argv[1]);
            std::string out = std::string(argv[2]);
            std::list<std::string> classes;
            for (int i = 3; i < argc; i++){
                classes.push_back(std::string(argv[i]));
            }
            std::cout << "\nThe result of the makefile is as follows\n\n";
            std::cout << generate_makefile(main, classes, out, "-g -Wall", "", true, true);
        }else{
            std::cout << "Error: Not give anyone name for .exe ";
        }
    }else{
        std::cout << "Creating Makefile for c++ with console interactive\n";
        std::list<std::string> classes;
        std::cout << "Name Main: ";
        std::string main;
        getline(std::cin, main);
        if (main.find(".cpp") != std::string::npos) {
            main = main.substr(0, main.length() - 4);
        }
        std::cout << "Name Main is " << main << std::endl;
        std::cout << "Name Out: ";
        std::string out;
        getline(std::cin, out);
        if (out.find(".exe") != std::string::npos) {
            out = out.substr(0, out.length() - 4);
        }
        std::string name;
        std::string answer;
        while(true){
            std::cout << "New class (Y/n): ";
            getline(std::cin, answer);
            if(answer == "n"){ break; }
            std::cout << "Name: ";
            getline(std::cin, name);
            if (name.find(".h") != std::string::npos) {
                name = name.substr(0, name.length() - 2);
            }
            classes.push_back(name);
        };
        std::string flags = "";
        std::string lflags = ""; // library flags
        
        std::cout << "Allow debugging (Y/n):";
        getline(std::cin, answer);
        if(answer != "n") flags += "-g "; 
        
        std::cout << "Show all warnings (Y/n):";
        getline(std::cin, answer);
        if(answer != "n") flags += "-Wall ";
        
        std::cout << "Include clean target (Y/n):";
        getline(std::cin, answer);
        bool clean = answer != "n";
        
        std::cout << "Include run target (Y/n):";
        getline(std::cin, answer);
        bool run = answer != "n";
        
        std::cout << "\nThe result of the makefile is as follows\n\n";
        std::cout << generate_makefile(main, classes, out, flags, lflags, clean, run);
    }
}

std::string help(){
    std::stringstream s;
    s << "[Generator of Makefiles for C++]\n";
    s << "Author: https://github.com/DavidVillalobos \n"; 
    s << "Usages:\n"; 
    s << "* Interactive console \n"; 
    s << "    makefile_generator \n"; 
    s << "* Through arguments \n";  
    s << "    makefile_generator [main] [out] [name_class_1] [name_class_2] ...";
    return s.str();
}

std::string displayList(std::list<std::string> &list, std::string separator){
    std::stringstream r;
    for (auto i : list){
        r << i << separator;
    }
    return r.str();
}

std::string compileTargetFiles(std::list<std::string> list){
    std::stringstream r;
    for (auto i : list){
        r << i + ".o: " + i + ".cpp\n\t$(CC) $(FLAGS) " + i + ".cpp\n\n";
    }
    return r.str();
}

std::string generate_makefile(std::string &main, std::list<std::string> &classes, 
std::string &out, std::string flags, std::string lflags, bool clean, bool run){
    std::stringstream makefile;
    makefile << "OBJS = " + main + ".o " + displayList(classes, ".o ") + '\n';
    makefile << "SOURCE = " + main + ".cpp " + displayList(classes, ".cpp ") + '\n';
    makefile << "HEADER = " + displayList(classes, ".h ") + '\n';
    makefile << "OUT = " + out + ".exe\n";
    makefile << "CC = g++\n";
    makefile << "FLAGS = -c " + flags + '\n';
    makefile << "LFLAGS = " + lflags + "\n\n"; 
    makefile << "all: $(OBJS)\n";
    makefile << "\t$(CC) -g $(OBJS) -o $(OUT) $(LFLAGS)\n\n";
    makefile << main + ".o: " + main + ".cpp\n";
    makefile << "\t$(CC) $(FLAGS) " + main + ".cpp\n\n";
    makefile << compileTargetFiles(classes);
    if(clean){
        makefile << "clean:\n\tdel $(OBJS) $(OUT)\n\n";
    } 
    if(run){
        makefile << "run: $(OUT)\n\t$(OUT)\n";
    } 
    std::string path = std::string(getenv("userprofile")) + "\\Desktop\\Makefile";
    std::ofstream file(path);
    file << makefile.str();
    file.close();
    return makefile.str();
}